ledger/acctupdates.go

```diff
@@ -655,11 +552,87 @@ func (au *accountUpdates) getCatchpointStream(round basics.Round) (io.ReadCloser
 
 // functions below this line are all internal functions
 
+// accountUpdatesLedgerEvaluator is a "ledger emulator" which is used *only* by initializeCaches, as a way to shortcut
+// the locks taken by the real ledger object when making requests that are being served by the accountUpdates.
+// Using this struct allow us to take the tracker lock *before* calling the loadFromDisk, and having the operation complete
+// without taking any locks. Note that it's not only the locks performance that is gained : by having the loadFrom disk
+// not requiring any external locks, we can safely take a trackers lock on the ledger during reloadLedger, which ensures
+// that even during catchpoint catchup mode switch, we're still correctly protected by a mutex.
+type accountUpdatesLedgerEvaluator struct {
```

Comment:
If I understand correctly, the whole point of this struct is so that `loadFromDisk` can take the `au.accountsMu` lock, right?

Reply:
That's true, but it extend further than that. The purpose of the struct is to allow bypassing the ledger's tracker mutex. Having the ability to cleanup the au.accountsMu is a bonus. ( which plays nicely ).

When using this struct, we can do the following:
1. make sure that on the Ledger's implementation, all code access to the `accountUpdates` are synchronized via the trackers mutex.
2. make sure that all the external method of  `accountUpdates` are all synchronized via the au.accountsMu ( including loadFromDisk )

#2 is particularly important during reloadLedger, since we want to take the Ledger's trackers mutex and then make changes to the `accountsUpdate` by calling `loadFromDisk`. If `loadFromDisk` requires to take the tracker lock during the eval, it would create a double-locking which won't work with current locking primitives.